### PR TITLE
Use existing Program class declaration in "Convert to 'Program.Main'" refactoring when possible

### PR DESCRIPTION
**Pull request by DoctorKrolic**
Fixes: https://github.com/dotnet/roslyn/issues/62943
Fixes: https://github.com/dotnet/roslyn/issues/62437

I think, @Youssef1313 thought, that refactoring reuses existing `Program` class declaration, but drops `static` modifier. It turned out, that refactoring wasn't using existing declaration at all, so when file contained suitable partial `Program` class it generated another partial `Program` class, making file content very messy. I made it reuse existing declaration and keep all available modifiers

### CODE REVIEW

**Review by Youssef1313 — COMMENTED** [review]
Actually I noticed now that there is originally no bug. The existing partial is kept so static is still there.

I'll leave it up to @CyrusNajmabadi to weigh on reusing the existing partial, it seems like a good improvement.

But consider adding a test where the partial program class is in a tree different than the top-level statements tree.

File: src/Features/CSharp/Portable/ConvertProgram/ConvertProgramTransform_ProgramMain.cs
Comment (Youssef1313) [inline src/Features/CSharp/Portable/ConvertProgram/ConvertProgramTransform_ProgramMain.cs:69]:
```suggestion
                   declaration.Modifiers.Any(SyntaxKind.PartialKeyword) &&
                   !declaration.Modifiers.Any(SyntaxKind.FileKeyword);
```

File: src/Features/CSharp/Portable/ConvertProgram/ConvertProgramTransform_ProgramMain.cs
Comment (Youssef1313) [inline src/Features/CSharp/Portable/ConvertProgram/ConvertProgramTransform_ProgramMain.cs:68]:
```suggestion
            return declaration.Identifier.ValueText == WellKnownMemberNames.TopLevelStatementsEntryPointTypeName &&
```

File: src/Features/CSharp/Portable/ConvertProgram/ConvertProgramTransform_ProgramMain.cs
Comment (Youssef1313) [inline src/Features/CSharp/Portable/ConvertProgram/ConvertProgramTransform_ProgramMain.cs:101]:
```suggestion
            ClassDeclarationSyntax? newClassDeclaration = null;
```

File: src/Features/CSharp/Portable/ConvertProgram/ConvertProgramTransform_ProgramMain.cs
Comment (Youssef1313) [inline src/Features/CSharp/Portable/ConvertProgram/ConvertProgramTransform_ProgramMain.cs:50]:
```suggestion
                            Contract.ThrowIfNull(newRoot);
                            newRoot = newRoot.RemoveNode(oldClassDeclaration, SyntaxGenerator.DefaultRemoveOptions);
```

File: src/Features/CSharp/Portable/ConvertProgram/ConvertProgramTransform_ProgramMain.cs
Comment (DoctorKrolic) [reply src/Features/CSharp/Portable/ConvertProgram/ConvertProgramTransform_ProgramMain.cs:101]: `newClassDeclaration` will be enforced to not be null later. If `oldClassDeclaration` is null, then we generate `newClassDeclaration` entirely, otherwise it will be taken as a modified `oldClassDeclaration` with new `Main` method member added. So this is kinda workaround to not bother with `newClassDeclaration` potentially being null when returning it.

Comment (Youssef1313) [reply src/Features/CSharp/Portable/ConvertProgram/ConvertProgramTransform_ProgramMain.cs:101]: @DoctorKrolic The compiler should be aware that `newClassDeclaration` is assigned to non-null and shouldn't produce warnings. Do you see any warnings if you use `ClassDeclarationSyntax? newClassDeclaration = null;`?

File: src/Features/CSharp/Portable/ConvertProgram/ConvertProgramTransform_ProgramMain.cs
Comment (DoctorKrolic) [reply src/Features/CSharp/Portable/ConvertProgram/ConvertProgramTransform_ProgramMain.cs:68]: Just curious, what is the difference between these 2 properties?

File: src/Features/CSharp/Portable/ConvertProgram/ConvertProgramTransform_ProgramMain.cs
Comment (DoctorKrolic) [reply src/Features/CSharp/Portable/ConvertProgram/ConvertProgramTransform_ProgramMain.cs:69]: Just checked, `file` modifier doesn't break program from compiling/running:
```cs
file class Program
{
    private static void Main(string[] args)
    {
        Console.WriteLine("Hello, World!");
    }
}
```
This will write `Hello, World!` to the console as expected

File: src/Features/CSharp/Portable/ConvertProgram/ConvertProgramTransform_ProgramMain.cs
Comment (Youssef1313) [reply src/Features/CSharp/Portable/ConvertProgram/ConvertProgramTransform_ProgramMain.cs:69]: @DoctorKrolic if you have `class @Program` for example, then `Text` is the actual source code text which is `@Program` while `ValueText` is just `Program`.

Also if you have `public class \u0050\u0072\u006F\u0067\u0072\u0061\u006D` (it might be surprising, but yes it's valid C# and is the same as `public class Program`)

File: src/Features/CSharp/Portable/ConvertProgram/ConvertProgramTransform_ProgramMain.cs
Comment (DoctorKrolic) [reply src/Features/CSharp/Portable/ConvertProgram/ConvertProgramTransform_ProgramMain.cs:102]: Verified it, compiler is really smart here, I am impressed) I thought, it would produce a warning

File: src/Features/CSharp/Portable/ConvertProgram/ConvertProgramTransform_ProgramMain.cs
Comment (DoctorKrolic) [reply src/Features/CSharp/Portable/ConvertProgram/ConvertProgramTransform_ProgramMain.cs:69]: 🤯

File: src/Features/CSharp/Portable/ConvertProgram/ConvertProgramTransform_ProgramMain.cs
Comment (Youssef1313) [reply src/Features/CSharp/Portable/ConvertProgram/ConvertProgramTransform_ProgramMain.cs:70]: @DoctorKrolic My bad. I thought that the following will generate two separate classes:

```csharp
System.Console.WriteLine();

file partial class Program { }
```

It might be worth adding a test anyway.

**Review by CyrusNajmabadi — APPROVED** [review]
nice!

File: src/Features/CSharp/Portable/ConvertProgram/ConvertProgramTransform_ProgramMain.cs
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/ConvertProgram/ConvertProgramTransform_ProgramMain.cs:124]: nit: no need for the `newClassDeclaration` local.  Can just return the items directly in the `if/else`.  that would be my preference.  but wasn't going to block on this.

**Pull request merged**
